1. 均方誤差 (MSE, Mean Squared Error)
- 公式： $MSE = \frac{1}{n}\sum(y - \hat{y})^2$  
- 定義： 誤差平方的平均值。  
- 單位： $單位^2$。  
- 懲罰： 強烈懲罰較大的錯誤 (Outliers)。  
- 用途： 常用作模型訓練時的損失函數 (Loss Function)。  
- 理想值： 越小越好 (最佳為 0)。  

2. 均方根誤差 (RMSE, Root Mean Squared Error)
- 公式： $RMSE = \sqrt{MSE} = \sqrt{\frac{1}{n}\sum(y - \hat{y})^2}$
- 定義： $MSE$ 的平方根。
- 單位： 與 $y$ 單位相同 (最易於解釋)。
- 懲罰： 懲罰較大的錯誤，但比 $MSE$ 緩和。
- 用途： 常用於最終報告，提供直觀的「平均誤差」大小。
- 理想值： 越小越好 (最佳為 0)。

3. 決定係數 ($R^2$, R-squared)
- 公式： $R^2 = 1 - \frac{\sum(y - \hat{y})^2}{\sum(y - \bar{y})^2}$
- 定義： 模型解釋的方差佔總方差的比例。
- 單位： 無單位 (標準化比例)。
- 度量： 相對度量，衡量模型相對於「只猜平均值」這個基準模型的改善程度。
- 值域： usually $0 \le R^2 \le 1$ (越高越好)。用途： 衡量模型的整體擬合優度 (Goodness of Fit)。

In [1]:
import numpy as np
from sklearn.metrics import mean_squared_error, r2_score
import math 
import pandas as pd 

In [2]:
# 真實值 (Actual Values)
y_true = np.array([10, 20, 30, 40])

# 模型 A 的預測 (Model A Predictions)
y_pred_A = np.array([11, 19, 32, 38])

# 模型 B 的預測 (Model B Predictions)
y_pred_B = np.array([10, 19, 31, 40])

# 創建一個 DataFrame 來總結數據，便於查看
data = {
    '真實值 (y_true)': y_true,
    '模型 A 預測 (y_pred_A)': y_pred_A,
    '模型 B 預測 (y_pred_B)': y_pred_B
}
df = pd.DataFrame(data)
print("--- 原始數據 ---")
print(df)
print("-" * 20)

--- 原始數據 ---
   真實值 (y_true)  模型 A 預測 (y_pred_A)  模型 B 預測 (y_pred_B)
0            10                  11                  10
1            20                  19                  19
2            30                  32                  31
3            40                  38                  40
--------------------


In [3]:
# --------------------
# 函數：計算所有指標
# --------------------
def evaluate_model(y_true, y_pred, model_name):
    print(f"\n===== 評估 {model_name} =====")
    
    # --- 1. 計算 MSE (均方誤差) ---
    # scikit-learn 的 mean_squared_error 默認 squared=True
    mse = mean_squared_error(y_true, y_pred)
    print(f"1. MSE (均方誤差): {mse:.3f}")
    
    # --- 2. 計算 RMSE (均方根誤差) ---
    # 方法一：手動開根號
    rmse_manual = np.sqrt(mse)
    # 方法二：使用 scikit-learn (在較新版本中可以直接設定 squared=False)
    # rmse_sklearn = mean_squared_error(y_true, y_pred, squared=False) 
    print(f"2. RMSE (均方根誤差): {rmse_manual:.3f}")
    
    # --- 3. 計算 R^2 (決定係數) ---
    r2 = r2_score(y_true, y_pred)
    print(f"3. R^2 (決定係數): {r2:.3f}")
    
    return [model_name, mse, rmse_manual, r2]

# --------------------
# 執行評估
# --------------------
results = []
results.append(evaluate_model(y_true, y_pred_A, "模型 A"))
results.append(evaluate_model(y_true, y_pred_B, "模型 B"))

# --------------------
# 總結比較
# --------------------
results_df = pd.DataFrame(results, columns=['模型', 'MSE', 'RMSE', 'R^2'])
print("\n" + "="*40)
print("             ** 總結比較 **")
print("="*40)
print(results_df)


===== 評估 模型 A =====
1. MSE (均方誤差): 2.500
2. RMSE (均方根誤差): 1.581
3. R^2 (決定係數): 0.980

===== 評估 模型 B =====
1. MSE (均方誤差): 0.500
2. RMSE (均方根誤差): 0.707
3. R^2 (決定係數): 0.996

             ** 總結比較 **
     模型  MSE      RMSE    R^2
0  模型 A  2.5  1.581139  0.980
1  模型 B  0.5  0.707107  0.996


- MSE：模型 B 的平方誤差平均值遠小於 A。
- RMSE：模型 B 的平均誤差約為 0.707，比 A 的 1.581 更接近真實值。單位與原始數據一致，最直觀。
- R2：模型 B 解釋了 99.6% 的數據方差，模型擬合程度極高，優於 A。